File: Backtracking/Permutations/permutations.py
class Solution:
      

    def permute(self, nums: List[int]) -> List[List[int]]:

        def backtrack(taken, sol, nums, res):
            if len(sol) == len(nums):
                res.append(sol.copy())
                return
            for i in range(len(nums)):
                if not taken[i]:
                    sol.append(nums[i])
                    taken[i] = True
                    backtrack(taken,sol,nums,res)
                    sol.pop()
                    taken[i] = False

        sol = []
        taken = [False] * len(nums)
        res = []
        backtrack(taken,sol,nums,res)
        return res

File: Backtracking/Permutations/permutations.cpp
class Solution {
public:

    void generateP(vector<int> &nums, vector<vector<int>> &res, int i){
        
        if(i == nums.size()){
            res.push_back(nums);
            return;
        }
        for(int j = i; j<nums.size(); j++){
            
            swap(nums[i],nums[j]);
            generateP(nums, res, i+1);
            swap(nums[j],nums[i]);
        }
    }

    vector<vector<int>> permute(vector<int>& nums) {

        vector<vector<int>> res;
        generateP(nums, res, 0);
        return res;
    }
};
